src/main/java/com/zendesk/maxwell/producer/AbstractAsyncProducer.java

```diff
@@ -0,0 +1,63 @@
+package com.zendesk.maxwell.producer;
+
+import com.zendesk.maxwell.replication.BinlogPosition;
+import com.zendesk.maxwell.MaxwellContext;
+import com.zendesk.maxwell.row.RowMap;
+
+public abstract class AbstractAsyncProducer extends AbstractProducer {
+	public class CallbackCompleter {
+		private InflightMessageList inflightMessages;
+		private final MaxwellContext context;
+		private final BinlogPosition position;
+		private final boolean isTXCommit;
+
+		public CallbackCompleter(InflightMessageList inflightMessages, BinlogPosition position, boolean isTXCommit, MaxwellContext context) {
+			this.inflightMessages = inflightMessages;
+			this.context = context;
+			this.position = position;
+			this.isTXCommit = isTXCommit;
+		}
+
+		public void markCompleted() {
+			if(isTXCommit) {
+				BinlogPosition newPosition = inflightMessages.completeMessage(position);
+
+				if(newPosition != null) {
+					context.setPosition(newPosition);
+				}
+			}
+		}
+	}
+
+	private InflightMessageList inflightMessages;
+
+	public AbstractAsyncProducer(MaxwellContext context) {
+		super(context);
+
+		this.inflightMessages = new InflightMessageList();
+	}
+
+	public abstract void push(RowMap r, CallbackCompleter cc) throws Exception;
```

Comment:
while we're naming stuff, how about this being called `sendAsync()` rather that it being polymorphic?